# Assignment 8: Sequence Composition

__Name:__

__Student ID:__


**Q1:** Can you see any obvious differences by eye?


**Q2:** Do the compositions of introns at different locations look
different by mono-nucleotide composition?

**Q3:** Do you see any obvious differences? Do you think these are
statistically relevant or just random noise? How could you tell the
difference?

**Q4:** Which introns are similar and different from others? Are first
introns different from introns farther downstream as suggested by the
biological model of IME?

**Q5:** The program reported some errors at high values of K. Any idea
why?

**Q6:** What is the average score of a proximal intron? What is the
average score of a distal intron? Use R to plot histograms. If you want
to get even more fancy, you could plot histograms for each intron
position (1, 2, 3, etc.).

**Q7:** What values of K work well? Why do you think that is? How does
data set size influence the optimal value of K?

**Q8:** Which intron in _A. thaliana_ is predicted to enhance expression
the most?

**Q9:** Extra credit questions. Supposing that the IME signal is a
protein binding site, how could you identify the binding site motif?
Given the motif, how could you determine if the signal was RNA or
DNA-based? How could you construct a synthetic intron more powerful than
any naturally occurring one?
