# Assignment 6 RNAseq PArt 2

Student Name: Amanda Everitt  

Student ID:998974934  

__Exercise 1__:  
__a.__ Use `merge()` to add gene descriptions for the genes found to be regulated by the DP treatment.  Output a table of the top 10 genes that includes the output from edgeR and the descriptions.  __Important: Pay attention to the "sort="" argument to `merge()`.  Should it be TRUE or FALSE?
```{r}
GeneDescription <- read.delim("./Brapa_reference/FileS9.txt", row.names = 1)
DEgene.interaction <- read.csv("Newgenes.gt.csv", row.names = 1) #the DEgenes.interaction.csv file
DEgenes.trt <- read.csv("DEgenes.trt.csv", row.names = 1)
DE.trt.merge <- merge(DEgenes.trt, GeneDescription, by="row.names", sort=FALSE) #because we don't want it sorted by the columns
head(DE.trt.merge, n=10)
```

__b.__ Repeat this for  genes with a genotype x trt interaction.
```{r}
DE.int.merge <- merge(DEgene.interaction, GeneDescription, by="row.names", sort=FALSE) #sort by rows, not columns
head(DE.int.merge, n=10 )
```

__Exercise 2__:  
```{r, include=FALSE}
library(goseq)
go.terms <- read.delim("./Brapa_reference/FileS11.txt",header=FALSE,as.is=TRUE)
head(go.terms)
names(go.terms) <- c("GeneID","GO")
summary(go.terms)

expressed.genes <- read.delim("internode_expressed_genes.txt",as.is=TRUE)
head(expressed.genes)
names(expressed.genes) <- "GeneID"

gene.lengths <- read.table("./Brapa_reference/Brapa_CDS_lengths.txt",as.is=TRUE)
head(gene.lengths)
summary(gene.lengths)

#we need to reduce the gene.length data to only contain entries for those genes in our expressed.genes set.  We also need this as a vector
gene.lengths.vector <- gene.lengths$Length[gene.lengths$GeneID %in% expressed.genes$GeneID]
names(gene.lengths.vector) <- gene.lengths$GeneID[gene.lengths$GeneID %in% expressed.genes$GeneID]
head(gene.lengths.vector)

#Do the reverse to make sure everything matches up (it seems that we don't have length info for some genes?)
expressed.genes.match <- expressed.genes[expressed.genes$GeneID %in% names(gene.lengths.vector),]
go.list <- strsplit(go.terms$GO,split=",")
names(go.list) <- go.terms$GeneID
head(go.list)
DE.interaction <- expressed.genes.match %in% rownames(DEgene.interaction) 
    #for each gene in expressed gene, return FALSE if it is not in DEgene.trt and TRUE if it is.
names(DE.interaction) <- expressed.genes.match
head(DE.interaction)
DE.interaction <- as.numeric(DE.interaction) #convert to 0s and 1s
head(DE.interaction)
sum(DE.interaction) # number of DE genes

#determines if there is bias due to gene length.  The plot shows the relationship.
nullp.result <- nullp(DEgenes = DE.interaction,bias.data = gene.lengths.vector)

#calculate p-values for each GO term
rownames(nullp.result) <- names(gene.lengths.vector) #because of a bug in nullp()
GO.out <- goseq(pwf = nullp.result,gene2cat = go.list,test.cats=("GO:BP"))

#list over-represented GO terms (p < 0.05)
#GO.out[GO.out$over_represented_pvalue < 0.05,]

#print(GO.out[GO.out$over_represented_pvalue < 0.05,1:2],row.names=FALSE)
```
__a:__ In REVIGO display a “TreeMap” of the BP GO terms. Was our hypothesis that cell wall genes are enriched in the genotype X treatment gene set correct? You DO NOT need to include the treemap in your answer.  

>Yes, the G2 cell wall genes compose 9% of genes the which have GO terms correlating to Biological Processes so preliminarily we can say that cell wall genes are enriched in our treatment. 

__b:__ Display a "TreeMap" of the CC GO terms.  There are four general categories shown, some with sub-categories.  What are the two general categories with the largest number of sub categories?  How might these general categories relate to differences in plant growth?  You DO NOT need to include the treemap in your answer.  

>The categories "Biological Process" and "Molecular Function" have the most sub categories. Generally, plants which have more genes in the GO category Biological Processes:transcription initiation may have more growth due to an increased promoter function. Similarly, plants which have more genes in the GO cateogry Molecular function:cytochrome receptor may have more growth because increased light harvesting. 

```{r, include=FALSE}
library(Biostrings) #R package for handling DNA and protein data
promoters <- readDNAStringSet("./Brapa_reference/BrapaV1.5_1000bp_upstream.fa.gz")

#convert "N" to "-" in promoters.  otherwise motifs will match strings of "N"s
promoters <- DNAStringSet(gsub("N","-",promoters))

promoters
motifs <- read.delim("./Brapa_reference/element_name_and_motif_IUPACsupp.txt",header=FALSE,as.is=TRUE)
head(motifs)
motifsV <- as.character(motifs[,2])
names(motifsV) <- motifs[,1]
motifsSS <- DNAStringSet(motifsV)
motifsSS
#get names to match...there are are few names in the DEgene list not in the promoter set
DEgene.interaction.match <- row.names(DEgene.interaction)[row.names(DEgene.interaction) %in% names(promoters)]

#subset promoter files
universe.promoters <- promoters[expressed.genes.match]
target.promoters <- promoters[DEgene.interaction.match]
#create a function to summarize the results and test for significance
motifEnrichment <- function(target.promoters,universe.promoters,all.counts=F,motifs=motifsSS) {
  
  #use vcountPDict to count the occurrences of each motif in each promoter
  target.counts <- vcountPDict(motifs,target.promoters,fixed=F) + 
    vcountPDict(motifsSS,reverseComplement(target.promoters),fixed=F)
  universe.counts <- vcountPDict(motifs,universe.promoters,fixed=F) + 
    vcountPDict(motifsSS,reverseComplement(universe.promoters),fixed=F)
  
  if (all.counts) { 
    #count all occurrences of a motif instead of the number of promoters that it occurs in
    target.counts.sum <- apply(target.counts,1,sum)
    universe.counts.sum <- apply(universe.counts,1,sum)
  } else {
    target.counts.sum <- apply(ifelse(target.counts > 0,1,0),1,sum)
    universe.counts.sum <- apply(ifelse(universe.counts > 0 , 1, 0),1,sum)
  }
  n.motifs <- length(target.counts.sum)
  results <- vector(mode="numeric",length=n.motifs)
  for (i in 1:n.motifs) {
    if (all.counts) { #the contigency tables are different depending on whether we are looking at promoters or overall occurrences
      #test if ratio of occurrences to promoters is the same in the target and the universe
      m <- matrix(c(
        target.counts.sum[i],                       #number of occurrences within target
        dim(target.counts)[2],                      #number of promoters in target
        universe.counts.sum[i],                  #number of occurrences within universe
        dim(universe.counts)[2]                  #number of promoters in universe
      ),ncol=2)
    } else { #looking at promoters with and without hits
      m <- matrix(c(
        target.counts.sum[i],                        #number of promoters in target with hit
        dim(target.counts)[2]-target.counts.sum[i],            #number of promoters in target with no hit
        universe.counts.sum[i],                   #number of promoters in universe with hit
        dim(universe.counts)[2]-universe.counts.sum[i]   #number of promoters in universe with no hit
      ),ncol=2)
    } #else
    results[i] <- fisher.test(m,alternative="greater")$p.value
  } #for loop
  results.table <- data.frame(
    motif=names(motifs),
    universe.percent = round(universe.counts.sum/dim(universe.counts)[2],3)*100,
    target.percent = round(target.counts.sum/dim(target.counts)[2],3)*100,
    p.value =  results)
  results.table <- results.table[order(results.table$p.value),]
  results.table
}
motif.results <- motifEnrichment(target.promoters,universe.promoters)
head(motif.results)
```

__Exercise 3__ :  
__a.__ How many motifs are enriched at P < 0.05?  

>There are six enriched motifs.
```{r}
enriched <- subset(motif.results, motif.results$p.value < 0.05)
enriched
```
__b.__ What is the identity of the most significantly over-enriched promoter?   

>The most enriched promoter, has the lowest p value :SORLIP1.  

__c.__ What percentage of genes in the "Universe" have this motif?  What percentage in our target set?  

>23.7% of the "Universe" have this motif and 25.1% of the target set have this motif.

__d.__ You can find information on the motifs
[here](http://arabidopsis.med.ohio-state.edu/AtcisDB/bindingsites.html).  Do you think that the most enriched motif represents a biologically meaningful result?  Discuss why or why not.  

>I belive this is a biologically meaningful result because it is a "key promoter motif involved in the network of light-regulated gene expression ". Given that this motif was enriched in a study with different light and dark conditions, this is a very promising motif of interest in our genes.