---
title: "Assignment 6 RNAseqPart 1"
output: html_document
---

__Score:__19

* **RD** -0.5 part2, Q3a. The answer was 7
* **RD** -0.5 part2, Q3b. The answer was SORLREP3 
* **RD** part2, Q3a. The answer was 8.2% (universal) 16.1% (target)

__Name:__ Amanda Everitt  
__Student ID:__ 998974934


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, include=FALSE}
library(rtracklayer)
gff <- import.gff("Brapa_reference/Brapa_gene_v1.5.gff")
gff #take a look

#create a column "gene_id" that contains the gene name for every entry
gff$gene_id <- ifelse(is.na(gff$ID),gff$Parent,gff$ID)

export(gff,"Brapa_reference/Brapa_gene_v1.5.gtf",format="gtf")

library(Rsubread)
readCounts <- featureCounts(
  files=c("../Assignment_5_Everitt.Amanda/tophat_out-IMB211_All_A01_INTERNODE.fq/accepted_hits_A01.bam",
    "../Assignment_5_Everitt.Amanda/tophat_out-R500_All_A01_INTERNODE.fq/accepted_hits_A01.bam"),
  annot.ext="Brapa_reference/Brapa_gene_v1.5.gtf", 
  isGTFAnnotationFile=TRUE,
  GTF.featureType="CDS", # This depends on GTF file.  Often it would be "exon"
  GTF.attrType="gene_id"
  )
```

__Exercise 1__  
Read the help file for feature counts.  Be sure to look at the section "Value" where it describes the output.  
__a__ Provide a line of code that displays the counts of the first 6 genes.  (It is not very interesting because the first genes in the file are on chromosome A03 (strange numbering...) and our bam file only has counts from A01...  )  
```{r}
justcounts <- as.data.frame(readCounts[1])
colnames(justcounts) <- c( "IMB", "R500")
head(justcounts, 6)
```

__b__ The gene `Bra011030` is on chromosome A01.  What are its read counts in the two files? 
```{r}
justcounts["Bra011030",]
```

__c__ What percentage of reads (from each file) were assigned to a gene?  What percentage were unassigned because they were not located in a gene (aka "Feature")?  

```{r}
readCounts$stat[1:4,] #show all lines which actually had values
```
>81.12% (1601807/1974510) of reads from IMB were assigned while 81.04% (1776266/2191853) reads from R500 were assigned. 10.87% (214694/1974510) of reads from IMB were assigned while 11.96% (262039/2191853) reads from R500 were assigned. 

__d__ What are 2 possible reasons why there are reads that cannot be assigned to a gene?  

>Reads may not be assigned to a gene because they have poor mapping quality or the fragment length is too small to map to any gene. 

__Exercise 2__  
Move the downloaded data to your current working directory.

__a__. Create a new object in R called `counts.data` with either the leaf data or the internode data.  
```{r}
counts.data <- read.delim("gh_internode_counts.tsv") #Internode data
```

__b__. Check to make sure that the data looks as expected.  (What do you expect and how do you confirm?  Show your commands and output.)

>I am expecting 12 samples, 40991 genes, DP, and NDP conditions.Looking in the global environment I can see there is 40991 obs. of 12 variables. Using summary I can see all of the conditions are included in the data frame.  
```{r}
#summary(counts.data)
counts.data[0,]
```

__Exercise 3__  
The column names are too long.  Use the `sub()` command to remove the ".1_matched.merged.fq.bam" suffix from each column name.  Although it doesn't matter in this case, using the argument "fixed=TRUE" is a good idea because "." is a wildcard character.
```{r}
colnames(counts.data) <- sub(".1_matched.merged.fq.bam", "", colnames(counts.data),fixed=TRUE)
names(counts.data)
```


__Exercise 4__  
__a.__ Make a histogram of counts for each of the samples.  
```{r}
counts.data <- counts.data[rownames(counts.data)!="*",]
counts.data[is.na(counts.data)] <- 0
library(reshape2)
# melt our data set to keep the column id.vars and place all the values from rep 2 and 3 at bottom 
#I know melt worked because the amount of obs in melt is 3x that of orignial data set
melt.data <- melt(counts.data)
library("ggplot2")
ggplot(melt.data, aes(value)) + geom_histogram() + facet_wrap(facets = ~ variable)
```

__b.__ Is the data normally distributed?  Apply an appropriate transformation if needed and make a new set of histograms.  
```{r}
ggplot(melt.data, aes(value)) + geom_histogram() + facet_wrap(facets = ~ variable) + scale_x_log10()
```

__Exercise 5:__  
We expect that read counts, especially from biological replicates, will be highly correlated.  Check to see if this is the case using the `pairs()` function and the `cor()` function.  Explain what each of these functions does and comment on the results.  __Important Hint:__ _`pairs` is slow on the full dataset.  Try it on the first 1,000 genes.  Do you need to transform to make the pairs output more meaningful?_  

```{r}
counts.data <- counts.data[rowSums(counts.data > 10) >= 3,]
cor(counts.data$IMB211_DP_1_INTERNODE, counts.data$IMB211_DP_2_INTERNODE)
cor(counts.data$R500_NDP_1_INTERNODE, counts.data$R500_NDP_3_INTERNODE)
cor(counts.data$IMB211_DP_1_INTERNODE, counts.data$R500_NDP_3_INTERNODE)
```
>Cor() computs the correlation of vectors x an y, here X and Y are our biological replicates. The closer the correlation coefficient is to one, it means the more positively correlated the two values are. Here we can see the biological replicates are more positively correlated to each other (0.88 and 0.99), than they are to different conditions (0.67).    

```{r}
counts.data.small <- counts.data[sample(nrow(counts.data),1000),] #random 1,000 genes
pairs(counts.data.small)
```

>Pairs() makes a matrix of scatter plots. This is a way to visualize multiple pairwise comparisons. 

```{r, include=FALSE}
sample.description <- data.frame(
  sample=colnames(counts.data),
  
  #This next line searches for IMB211 or R500 in the colnames of counts.data and returns anything that matches
  #In this way we can extract the genotype info.
  gt=regmatches(colnames(counts.data),regexpr("R500|IMB211",colnames(counts.data))),
  
  #Now we use the same method to get the treatment
  trt=regmatches(colnames(counts.data),regexpr("NDP|DP",colnames(counts.data)))
  )

# Now we can paste the trt and gt columns together to give a group identifier
sample.description$group <- paste(sample.description$gt,sample.description$trt,sep="_")

# set the reference treatment to "NDP"
sample.description$trt <- relevel(sample.description$trt,ref="NDP")

sample.description

library(edgeR)
dge.data <- DGEList(counts=counts.data, group=sample.description$group)
dim(dge.data) 
dge.data <- calcNormFactors(dge.data, method = "TMM")
dge.data$samples # look at the normalization factors
```

__Exercise 6__  
Discuss the MDS plot.  Does it give you confidence in the experiment or cause concern?       

```{r}
plotMDS(dge.data, method = "bcv")
```

>This plot gives more confidence in the experiment because it shows clear seperations of the four conditions. The BCV distance is less between replicates that in it between conditions.Specifically, BCV1 seperated by R500 and IMB211 whereas BCV2 seperated by DP and NDP.

__Exercise 7__  
To get a graphical idea for what the normalization does, make box plots of the count data for each sample before and after normalization.  Discuss the effect of normalization.   
```{r}
counts.data.normal <- cpm(dge.data)
counts.data.normal.log <- log2(counts.data.normal +1) # for normalized data
counts.data.log <- log2(counts.data +1)   #for untransformed data
melt.data.normal <- melt(counts.data.normal.log)
boxplot(counts.data.normal.log)
boxplot(counts.data.log)
```

>The normalized boxplots have much more consistent medians and quartiles than the non-normalized. They all fall around 4 versus in the normalized versus falling between 3 and 8 in the non-normalized.  

__Exercise 8__  
```{r, include=FALSE}
design <- model.matrix(~gt+trt,data = sample.description)
rownames(design) <- sample.description$sample
design
#First the overall dispersion
dge.data <- estimateGLMCommonDisp(dge.data,design,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data <- estimateGLMTrendedDisp(dge.data,design)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data <- estimateGLMTagwiseDisp(dge.data,design)

#We can examine this with a plot
plotBCV(dge.data)

fit <- glmFit(dge.data, design)
gt.lrt <- glmLRT(fit,coef = "gtR500")
topTags(gt.lrt) # the top 10 most differentially expressed genes
summary(decideTestsDGE(gt.lrt,p.value=0.01)) #This uses the FDR.  0.05 would be OK also.
#Extract genes with a FDR < 0.01 (could also use 0.05)
DEgene.gt <- topTags(gt.lrt,n = Inf)$table[topTags(gt.lrt,n = Inf)$table$FDR<0.01,]

#save to a file
write.csv(DEgene.gt,"DEgenes.gt.csv")
```

```{r, include=FALSE}
plotDE <- function(genes, dge, sample.description) {
  require(ggplot2)
  require(reshape2)
  tmp.data <- t(log2(cpm(dge[genes,])+1))
  tmp.data <- merge(tmp.data,sample.description,by.x="row.names",by.y="sample")
  tmp.data <- melt(tmp.data,value.name="log2_cpm",variable.name="gene")
  pl <- ggplot(tmp.data,aes(x=gt,y=log2_cpm,fill=trt))
  pl <- pl + facet_wrap( ~ gene)
  pl <- pl + ylab("log2(cpm)") + xlab("genotype")
  pl <- pl + geom_boxplot()
  pl + theme(axis.text.x  = element_text(angle=45, vjust=1,hjust=1))
}
# A single gene
plotDE("Bra009785",dge.data,sample.description)

#top 9 genes
plotDE(rownames(DEgene.gt)[1:9],dge.data,sample.description)
```
 
 _a_.  Find all genes differentially expressed in response to the DP treatment (at a FDR < 0.01).
```{r}
trt.lrt <- glmLRT(fit,coef = "trtDP")
topTags(trt.lrt)
```

_b_.  How many genes are differentially expressed?  
```{r}
summary(decideTestsDGE(trt.lrt,p.value=0.01))
```
>618(199+419) genes are differentially expressed. 

_c_.  Make a plot of the top 9
```{r}
DEgene.trt <- topTags(trt.lrt,n = Inf)$table[topTags(trt.lrt,n = Inf)$table$FDR<0.01,]
write.csv(DEgene.trt,"DEgenes.trt.csv")
plotDE(rownames(DEgene.trt)[1:9],dge.data,sample.description)
```

__Exercise 9__:  

__a__. Repeat the dispersion estimates and model fit but with the new model.  Show code.  
```{r}
design.interaction <- model.matrix(~gt*trt,data = sample.description)
rownames(design.interaction) <- sample.description$sample
#design.interaction

dge.data.new <- estimateGLMCommonDisp(dge.data,design.interaction,verbose = TRUE)

#Then a trended dispersion based on count level
dge.data.new <- estimateGLMTrendedDisp(dge.data.new,design.interaction)

#And lastly we calculate the gene-wise dispersion, using the prior estimates to "squeeze" the dispersion towards the common dispersion.
dge.data.new <- estimateGLMTagwiseDisp(dge.data.new,design.interaction)

#We can examine this with a plot
plotBCV(dge.data.new)

fit.new <- glmFit(dge.data.new, design)
gt.lrt.new <- glmLRT(fit,coef = "gtR500")
#summary(decideTestsDGE(gt.lrt.new, p.value=0.01)) #This uses the FDR.  0.05 would be OK also.
Newgene.gt <- topTags(gt.lrt.new,n = Inf)$table[topTags(gt.lrt.new,n = Inf)$table$FDR<0.01,]

#save to a file
write.csv(Newgene.gt,"Newgenes.gt.csv")
```

__b__. How many genes show a significantly different response to treatment in IMB211 as compared to R500?  Save these genes to a file.  
```{r}
#topTags(gt.lrt.new) # the top 10 most differentially expressed genes
summary(decideTestsDGE(gt.lrt.new,p.value=0.01)) #Extract genes with a FDR < 0.01 
```
>10824 (5667+5157) genes show a significant response to treatment. 

__c__. Make a plot of the top 9 genes that have a significantly different response to treatment in IMB211 as compared to R500.  
```{r}
plotDE(rownames(Newgene.gt)[1:9],dge.data.new,sample.description)
```

